src/violation.rs

```diff
@@ -1,15 +1,98 @@
-// TODO: Implement Violations for constraints
-#[derive(Debug, Clone)]
-pub struct Violations {
-    violations: Vec<Violation>,
+/// Represents Violations found during validation of values based on schema types
+pub trait Violations {
+    /// adds a violation into the collection of [Violation]s
+    fn add_violation(&mut self, violation: ViolationImpl);
+
+    /// Returns `true` if no violations were found, otherwise `false`
+    fn is_valid(&self) -> bool;
+
+    /// Returns the [Violation]s that were found during validation
+    fn violations(&self) -> &[ViolationImpl];
+
+    /// Creates a checkpoint to track the number of violations from this point onwards
+    fn checkpoint(&self) -> Checkpoint;
+}
+
+// Represents collection of [Violation]s
+pub struct ViolationsImpl {
+    violations: Vec<ViolationImpl>,
+}
+
+impl ViolationsImpl {
+    pub fn new() -> Self {
+        Self { violations: vec![] }
+    }
 }
 
-impl Violations {
-    fn add_violation(violation: Violation) {
-        todo!()
+impl Violations for ViolationsImpl {
+    fn add_violation(&mut self, violation: ViolationImpl) {
+        self.violations.push(violation);
     }
+
+    fn is_valid(&self) -> bool {
+        self.violations.is_empty()
+    }
+
+    fn violations(&self) -> &[ViolationImpl] {
+        &self.violations
+    }
+
+    fn checkpoint(&self) -> Checkpoint {
+        Checkpoint::new(self.violations.len())
+    }
+}
+
+// TODO: this can be extended to store children violation count as well for container constraints
+// Represents a checkpoint to be used while validation
+// This determines if children or nested types/constraints have the same number of violations since the checkpoint was created
+pub struct Checkpoint {
```

Comment:
Is this how `ion-schema-kotlin` does this? I'm wondering what the complexity will buy us.

Reply:
Yes, this is what `ion-schema-kotlin` does, but it's only needed because `ion-schema-kotlin` relies on side-effects to produce the complete tree of violations. The logical constraints `one_of`, `any_of`, and `all_of` use the check point to see if there are any violations for each type that they test. 

Simplified version of what `validate` does in the logical constraints:
```
val validTypes = mutableListOf<Type>()
types.forEach { t ->
    val checkpoint = violation.checkpoint()
    t.validate(value, violation)
    if (checkpoint.isValid()) {
        validTypes.add(it())
    }
}
// And then more logic to handle checking the number of matching types in `validTypes`
```
The `Checkpoint` type is a tuple of the sizes of `violations` and `children` of the current `Violation`. If there was only a single list of child objects (instead of `violations` and `children`) we could replace `Checkpoint` with an `Int`.

If we could do could make `validate` a pure function, it would be even better:
```
// Assume we have a kotlin sealed class like Rust Option type.
// typeResults is inferred to be Map<Type, Option<Violation>>

val typeResults = types.associateWith { t -> t.validate(value) }
val validTypes = typeResults.filterValues { it is Option.None }.keys
```

IMO, the complexity of `Checkpoint` is not worth it.

Reply:
Removed `Checkpoint` struct using the pure validate function which return `Option<Violation>`.